leaderelection/leaderelection.go

```diff
@@ -0,0 +1,154 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Config is put into a different package to prevent cyclic imports in case
+// it is needed in several locations
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"github.com/elastic/cloudbeat/config"
+	"github.com/elastic/cloudbeat/resources/providers"
+	"github.com/elastic/elastic-agent-autodiscover/kubernetes"
+	"github.com/elastic/elastic-agent-libs/logp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	le "k8s.io/client-go/tools/leaderelection"
+	rl "k8s.io/client-go/tools/leaderelection/resourcelock"
+	"os"
+	"strings"
+	"sync"
+)
+
+var (
+	callOnce sync.Once
+)
+
+const ()
+
+type ElectionManager interface {
+	Run(onNewLeader func(ctx context.Context) error) error
+	IsLeader() bool
+}
+
+type Manager struct {
+	log    *logp.Logger
+	ctx    context.Context
+	client k8s.Interface
+	leader *le.LeaderElector
+}
+
+func NewLeaderElector(ctx context.Context, log *logp.Logger, cfg config.Config) (ElectionManager, error) {
+	kubeClient, err := providers.KubernetesProvider{}.GetClient(cfg.KubeConfig, kubernetes.KubeClientOptions{})
+	if err != nil {
+		log.Errorf("NewLeaderElector error in GetClient: %v", err)
+		return nil, err
+	}
+
+	return &Manager{
+		ctx:    ctx,
+		log:    log,
+		client: kubeClient,
+		leader: nil,
+	}, nil
+}
+
+// Run runs leader election and calls onStarted when lease has been acquired
+func (m *Manager) Run(onNewLeader func(ctx context.Context) error) error {
+	var err error
+	leaderElectionConf := m.getConfig(onNewLeader)
+
+	m.leader, err = le.NewLeaderElector(leaderElectionConf)
+	if err != nil {
+		return err
+	}
+
+	go m.leader.Run(m.ctx)
+	m.log.Infof("started leader election", "for", leaderElectionConf.Lock.Describe(), "id", leaderElectionConf.Lock.Identity())
+
+	return nil
+}
+
+func (m *Manager) IsLeader() bool {
+	return m.leader.IsLeader()
```

Comment:
so will this work properly what the manages is shared by many gorutines?
I think it will.
previous version seems to be not concurrent safe.